tensorflow/compiler/mlir/xla/transforms/tensor_linalg_to_buffer_linalg.cc

```diff
@@ -0,0 +1,152 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// This file implements logic for transforming Linalg operations with tensor
+// types to memref type and allocate and deallocate buffers using the
+// BufferAssignmentPlacer.
+
+#include "absl/memory/memory.h"
+#include "llvm/ADT/APInt.h"
+#include "mlir/Dialect/Linalg/IR/LinalgOps.h"    // TF:llvm-project
+#include "mlir/Dialect/Linalg/IR/LinalgTypes.h"  // TF:llvm-project
+#include "mlir/Dialect/StandardOps/IR/Ops.h"     // TF:llvm-project
+#include "mlir/IR/AffineExpr.h"                  // TF:llvm-project
+#include "mlir/IR/Attributes.h"                  // TF:llvm-project
+#include "mlir/IR/Builders.h"                    // TF:llvm-project
+#include "mlir/IR/Function.h"                    // TF:llvm-project
+#include "mlir/IR/Location.h"                    // TF:llvm-project
+#include "mlir/IR/MLIRContext.h"                 // TF:llvm-project
+#include "mlir/IR/Operation.h"                   // TF:llvm-project
+#include "mlir/IR/PatternMatch.h"                // TF:llvm-project
+#include "mlir/IR/StandardTypes.h"               // TF:llvm-project
+#include "mlir/Pass/Pass.h"                      // TF:llvm-project
+#include "mlir/Transforms/DialectConversion.h"   // TF:llvm-project
+#include "tensorflow/compiler/mlir/xla/ir/lhlo_ops.h"
+#include "tensorflow/compiler/mlir/xla/transforms/buffer_assignment.h"
+#include "tensorflow/compiler/mlir/xla/transforms/rewriters.h"
+
+namespace mlir {
+namespace xla {
+namespace {
+
+class GenericOpConverter
+    : public xla::BufferAssignmentOpConversionPattern<linalg::GenericOp> {
+ public:
+  using xla::BufferAssignmentOpConversionPattern<
+      linalg::GenericOp>::BufferAssignmentOpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      linalg::GenericOp op, ArrayRef<Value> operands,
+      ConversionPatternRewriter& rewriter) const final {
+    auto loc = op.getLoc();
+    SmallVector<Value, 4> args(operands.begin(), operands.end());
+
+    // Update all types to memref types.
+    auto results = op.getOperation()->getResults();
+    for (auto result : results) {
+      auto type = result.getType().cast<ShapedType>();
+      if (!type)
+        op.emitOpError()
+            << "tensor to buffer conversion expects ranked results";
```

Comment:
maybe "expects shaped results"?  Also this chekc is probably not needed at all, since Linalg verifiers check that: 
in LinalgStructuredOps.td:

```
def LinalgOperand: Type<
  Or<[AnyRankedTensor.predicate, AnyStridedMemRef.predicate]>>;

class LinalgOperandOfRank<int rank>: Type<
  And<[
    LinalgOperand.predicate,
    CPred<"$_self.cast<ShapedType>().getRank() == " # rank>]
  >>;

class GenericOpBase<string mnemonic> : LinalgStructuredBase_Op<mnemonic, []> {
  let arguments = (ins Variadic<LinalgOperand>:$views,
                   I64Attr:$args_in,
                   I64Attr:$args_out,
```